x/sync/manager.go

```diff
@@ -234,38 +283,59 @@ func (m *Manager) close() {
 	})
 }
 
-// Processes [item] by fetching and applying a change or range proof.
-// Assumes [m.workLock] is not held.
+func (m *Manager) finishWorkItem() {
+	m.workLock.Lock()
+	defer m.workLock.Unlock()
+
+	m.processingWorkItems--
+	m.unprocessedWorkCond.Signal()
+}
+
+// Processes [item] by fetching a change or range proof.
 func (m *Manager) doWork(ctx context.Context, work *workItem) {
-	defer func() {
-		m.workLock.Lock()
-		defer m.workLock.Unlock()
+	if errors.Is(ctx.Err(), context.Canceled) {
+		m.finishWorkItem()
+		return
+	}
 
-		m.processingWorkItems--
-		m.unprocessedWorkCond.Signal()
-	}()
+	// Backoff for failed requests accounting for time this job has already
+	// spent waiting in the unprocessed queue
+	now := time.Now()
+	waitTime := max(0, calculateBackoff(work.attempt)-now.Sub(work.queueTime))
+
+	// Check if we can start this work item before the context deadline
+	deadline, ok := ctx.Deadline()
+	if ok && now.Add(waitTime).After(deadline) {
+		m.finishWorkItem()
+		return
+	}
+
+	<-time.After(waitTime)
```

Comment:
I think we are still blocking here if the context is cancelled. I think we should remove the above check for `ctx.Err()` and use a `select`